File: src/main/kotlin/org/kite9/diagram/visualization/orthogonalization/vertex/MultiElementVertexArranger.java
package org.kite9.diagram.visualization.orthogonalization.vertex;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.kite9.diagram.common.elements.mapping.ElementMapper;
import org.kite9.diagram.common.elements.vertex.ContainerSideVertex;
import org.kite9.diagram.common.elements.vertex.DartJunctionVertex;
import org.kite9.diagram.common.elements.vertex.EdgeCrossingVertex;
import org.kite9.diagram.common.elements.vertex.MultiCornerVertex;
import org.kite9.diagram.common.elements.vertex.MultiElementVertex;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.common.objects.OPair;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.orthogonalization.DartFace;
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization;
import org.kite9.diagram.visualization.orthogonalization.edge.IncidentDart;
import org.kite9.diagram.visualization.planarization.mgt.BorderEdge;
import org.kite9.diagram.logging.LogicException;

/**
 * Creates darts for edges arriving at corners or sides of a planarization.
 * 
 * It is expected that multiple edges could arrive to a single point in the same direction.  This should split those out into 
 * some side-edge arrangement.
 * 
 * @author robmoffat
 *
 */
public class MultiElementVertexArranger extends ConnectedVertexArranger {
	
	public MultiElementVertexArranger(ElementMapper em) {
		super(em);
	}
	
	/**
	 * This exclude {@link EdgeCrossingVertex}s and grid elements.
	 */
	@Override
	public boolean needsConversion(Vertex v) {
		if (((v instanceof MultiCornerVertex) && (((MultiCornerVertex)v).getDiagramElements().size()==1)) 
				|| (v instanceof ContainerSideVertex)) {
			return true;
		} else {
			return super.needsConversion(v);
		}
	}
	
	@Override
	protected DartFace convertVertex(Orthogonalization o, Vertex v, TurnInformation ti) {
		if (v instanceof MultiElementVertex) {
			Map<Direction, List<IncidentDart>> dartDirections = getDartsInDirection(v, o, ti);
			List<IncidentDart> allIncidentDarts = convertToDartList(dartDirections);
			createBorderEdges(allIncidentDarts, dartDirections, o, (MultiElementVertex) v);
			log.send("Converting vertex: "+v, allIncidentDarts);
			setupBoundariesFromIncidentDarts(allIncidentDarts, v);
			return null;
		} else {
			return super.convertVertex(o, v, ti);
		}
	}

	private void createBorderEdges(List<IncidentDart> dartDirections, Map<Direction, List<IncidentDart>> map, Orthogonalization o, MultiElementVertex v) {
		// create sides for any sides that have > 1 incident darts
		OPair<IncidentDart> borders = identifyBorders(dartDirections);
		Direction inDirection = Direction.reverse(borders.getA().getArrivalSide());
		Direction outDirection = borders.getB().getArrivalSide();
		
		if (inDirection == outDirection) {
			// single side
			Direction sideDirection = Direction.rotateClockwise(inDirection);
			
			Map<DiagramElement, Direction> aUnderlying = ((BorderEdge) borders.getA().getDueTo()).getDiagramElements();
			Map<DiagramElement, Direction> bUnderlying = ((BorderEdge) borders.getB().getDueTo()).getDiagramElements();
			
			if ((!aUnderlying.equals(bUnderlying)) || (aUnderlying == null)) {
				throw new LogicException();
			}

			List<IncidentDart> dartsToUse = map.get(sideDirection);
			if (dartsToUse.size() == 0) {
				sideDirection = Direction.reverse(sideDirection);
				dartsToUse = map.get(sideDirection);
			}
			
			createSide(borders.getA().getInternal(), borders.getB().getInternal(), dartsToUse, o, inDirection, aUnderlying);
		} else {
			Vertex midPoint = new DartJunctionVertex("mcv-"+newVertexId++, v.getDiagramElements());
			
			Direction aSide = Direction.reverse(outDirection);
			Direction bSide = inDirection;
			DiagramElement aUnderlying = ((BorderEdge) borders.getA().getDueTo()).getElementForSide(aSide);
			DiagramElement bUnderlying = ((BorderEdge) borders.getB().getDueTo()).getElementForSide(bSide);

			if (aUnderlying != bUnderlying) {
				throw new LogicException();
			}

			Map<DiagramElement, Direction> aUnderlyings = ((BorderEdge) borders.getA().getDueTo()).getDiagramElements();
			Map<DiagramElement, Direction> bUnderlyings = ((BorderEdge) borders.getB().getDueTo()).getDiagramElements();
			
			List<IncidentDart> side1Darts = map.get(aSide);
			List<IncidentDart> side2Darts = map.get(inDirection);
			
			
			createSide(borders.getA().getInternal(), midPoint, side1Darts, o, inDirection, aUnderlyings);
			createSide(midPoint, borders.getB().getInternal(), side2Darts, o, outDirection, bUnderlyings);
		}
	}

	private OPair<IncidentDart> identifyBorders(List<IncidentDart> dartDirections) {
		// there should be two IncidentDarts next to each other that represent the edge of the container
		List<IncidentDart> borders = dartDirections.stream().filter(d -> d.getDueTo() instanceof BorderEdge).collect(Collectors.toList());
		
		if (borders.size() != 2) {
			throw new LogicException();
		}
		
		IncidentDart to = borders.get(0);
		IncidentDart from = borders.get(1);

		int idx1 = dartDirections.indexOf(to);
		int idx2 = dartDirections.indexOf(from);
		
		if (idx2 - idx1 == 1) {
			return new OPair<IncidentDart>(from, to);
		} else if ((idx1 == 0) && (idx2 == dartDirections.size()-1)) {
			return new OPair<IncidentDart>(to, from);
		} else {
			throw new LogicException();
		}
	}
	
}

File: kite9-visualization/src/common/kotlin/org/kite9/diagram/visualization/orthogonalization/vertex/MultiElementVertexArranger.kt
package org.kite9.diagram.visualization.orthogonalization.vertex

import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.mapping.ElementMapper
import org.kite9.diagram.common.elements.vertex.*
import org.kite9.diagram.common.objects.OPair
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.Connection
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Direction.Companion.reverse
import org.kite9.diagram.model.position.Direction.Companion.rotateClockwise
import org.kite9.diagram.visualization.orthogonalization.DartFace
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization
import org.kite9.diagram.visualization.orthogonalization.edge.IncidentDart
import org.kite9.diagram.visualization.orthogonalization.vertex.VertexArranger.TurnInformation
import org.kite9.diagram.visualization.planarization.Tools
import org.kite9.diagram.visualization.planarization.mgt.BorderEdge

/**
 * Creates darts for edges arriving at corners or sides of a planarization.
 *
 * It is expected that multiple edges could arrive to a single point in the same direction.  This should split those out into
 * some side-edge arrangement.
 *
 * @author robmoffat
 */
open class MultiElementVertexArranger(em: ElementMapper) : ConnectedVertexArranger(em) {
    
    /**
     * This exclude [EdgeCrossingVertex]s and grid elements.
     */
    override fun needsConversion(v: Vertex): Boolean {
        return if ((v is MultiCornerVertex && v.getDiagramElements().size == 1)
            || (v is EdgeCrossingVertex && v.getDiagramElements().size == 1)
            || v is ContainerSideVertex
        ) {
            true
        } else {
            super.needsConversion(v)
        }
    }

    override fun convertVertex(o: Orthogonalization, v: Vertex, ti: TurnInformation): DartFace? {
        return if (v is MultiElementVertex) {
            val dartDirections: Map<Direction, List<IncidentDart>> =
                getDartsInDirection(v, o, ti)
            val allIncidentDarts: List<IncidentDart> = convertToDartList(dartDirections)
            createBorderEdges(allIncidentDarts, dartDirections, o, v)
            log.send("Converting vertex: $v", allIncidentDarts)
            setupBoundariesFromIncidentDarts(allIncidentDarts, v)
            null
        } else {
            super.convertVertex(o, v, ti)
        }
    }

    private fun createBorderEdges(
        dartDirections: List<IncidentDart>,
        map: Map<Direction, List<IncidentDart>>,
        o: Orthogonalization,
        v: MultiElementVertex
    ) {
        // create sides for any sides that have > 1 incident darts
        val (a, b) = identifyBorders(dartDirections)
        val inDirection = reverse(a.arrivalSide)
        val outDirection = b.arrivalSide
        if (inDirection === outDirection) {
            // single side
            if (a.internal != b.internal) {
                var sideDirection: Direction? = rotateClockwise(inDirection)
                val aUnderlying = (a.dueTo as BorderEdge).getDiagramElements() as Map<DiagramElement, Direction>
                val bUnderlying = (b.dueTo as BorderEdge).getDiagramElements() as Map<DiagramElement, Direction>
                if (aUnderlying != bUnderlying || aUnderlying == null) {
                    throw LogicException()
                }
                var dartsToUse = map[sideDirection]!!
                if (dartsToUse.size == 0) {
                    sideDirection = reverse(sideDirection)
                    dartsToUse = map[sideDirection]!!
                }
                createSide(a.internal, b.internal, dartsToUse, o, inDirection, aUnderlying)
            }
        } else {
            val midPoint: Vertex = DartJunctionVertex("mcv-" + newVertexId++, v.getDiagramElements())
            val aSide = reverse(outDirection)
            val aUnderlying = (a.dueTo as BorderEdge).getElementForSide(aSide!!)
            val bUnderlying = (b.dueTo as BorderEdge).getElementForSide(inDirection!!)
            if (aUnderlying !== bUnderlying) {
                throw LogicException()
            }
            val aUnderlyings = a.dueTo.getDiagramElements() as Map<DiagramElement, Direction>
            val bUnderlyings = b.dueTo.getDiagramElements() as Map<DiagramElement, Direction>
            val side1Darts = map[aSide]!!
            val side2Darts = map[inDirection]!!
            createSide(a.internal, midPoint, side1Darts, o, inDirection, aUnderlyings)
            createSide(midPoint, b.internal, side2Darts, o, outDirection, bUnderlyings)
        }
    }

    private fun identifyBorders(dartDirections: List<IncidentDart>): OPair<IncidentDart> {
        // there should be two IncidentDarts next to each other that represent the edge of the container
        val borders = dartDirections.filter { d -> d.dueTo is BorderEdge }

        if (borders.size != 2) {
            throw LogicException()
        }
        val to = borders[0]
        val from = borders[1]
        val idx1 = dartDirections.indexOf(to)
        val idx2 = dartDirections.indexOf(from)
        return if (idx2 - idx1 == 1) {
            OPair(from, to)
        } else if (idx1 == 0 && idx2 == dartDirections.size - 1) {
            OPair(to, from)
        } else {
            throw LogicException()
        }
    }

    private fun createPortTreeExternalVertex(
        e: PortVertex,
        list: List<PlanarizationEdge>
    ): ExternalVertex {
        return ExternalVertex(e.getID() + "-pve" + newVertexId++, list.toSet(), true)
    }

    override fun handleEdgeBucketing(
        ti: TurnInformation,
        cd: Set<DiagramElement>,
        o: Orthogonalization,
        from: Vertex,
        list: List<PlanarizationEdge>
    ): List<IncidentDart> {
        var incidentDirection : Direction = Direction.DOWN
        if ((from is PortVertex) && (list.size > 1)) {
            // in this case we need to _not_ separate out elements into different side vertices.
            var sideVertex = createSideVertex(cd, from)
            val portTreeVertex: Vertex = createPortTreeExternalVertex(from, list)

            val outList = mutableListOf<IncidentDart>()
            var lastStartVertex : Vertex? = null
            val midPoint : Float = calculateMidPoint(list)
            for (i in list.indices) {
                val current = list[i]
                incidentDirection = ti.getIncidentDartDirection(current)
                var startVertex = if (i.toFloat() == midPoint) portTreeVertex else createExternalVertex(current, from)
                var id = ec.convertPlanarizationEdge(current, o, incidentDirection, createExternalVertex(current, from), startVertex, from, null)

                outList.add(id)

                if (lastStartVertex != null) {
                    val subList = if ( i <= midPoint) {
                        list.subList(0, i)
                    } else {
                        list.subList(i, list.size)
                    }

                    val underlyings = subList
                        .flatMap { it.getDiagramElements().keys }
                        .map { it to null }
                        .toMap()

                    o.createDart(lastStartVertex, startVertex, underlyings, Direction.rotateAntiClockwise(incidentDirection))
                }

                lastStartVertex = startVertex

                if ((i < midPoint) && (i + 1 > midPoint)) {
                    // join fork up with portTreeVertex
                    val subList = list.subList(0, i+1)
                    val underlyings = subList
                        .flatMap { it.getDiagramElements().keys }
                        .map { it to null }
                        .toMap()

                    o.createDart(startVertex, portTreeVertex, underlyings, Direction.rotateAntiClockwise(incidentDirection))
                    lastStartVertex = portTreeVertex
                }
            }

            // join the portTree to the side
            o.createDart(portTreeVertex, sideVertex,
                list
                    .flatMap { it.getDiagramElements().keys }
                    .map { it to null}
                    .toMap(),
                incidentDirection
            )

            // rewrite incident darts using side
            return outList.map {
                IncidentDart(it.external, sideVertex, it.arrivalSide, it.dueTo)
            }
        } else {
            return super.handleEdgeBucketing(ti, cd, o, from, list)
        }
    }

    private fun calculateMidPoint(list: List<PlanarizationEdge>) : Float {
        // if we have a straight edge, use that
        val straightEdges = list
            .filter {
                it.isStraightInPlanarization() &&
                !Tools.isUnderlyingContradicting(it)
            }

        if (straightEdges.size > 1) {
            throw LogicException("Should only be one non-contradicting straight edge leaving a port: $straightEdges")
        } else if (straightEdges.isEmpty()) {
            return (list.size - 1.0f) / 2.0f
        } else {
            return list.indexOf(straightEdges.first()).toFloat()
        }
    }

}